File: tests/cpp/test_doom_automap.cpp
// examples/doom/automap.md 검증 (C++17).
#include "doom_automap.hpp"
#include "minitest.hpp"

using namespace Doom;
using namespace Doom::Map;

static Seg makeSegWithSecret(Vertex& v1, Vertex& v2, Sector& sector, SideDef& side, LineDef& line, bool secret) {
    side.sector = &sector;
    line = LineDef{&v1, &v2, secret ? uint16_t(LineDef::SECRET_FLAG) : uint16_t(0), 0, 0, &side, nullptr};
    return Seg{&v1, &v2, 0, &line, 0, 0};
}

TEST(automap_toggle_flips_open_state) {
    AutoMap m;
    CHECK(!m.isOpen());
    m.toggle();
    CHECK(m.isOpen());
    m.toggle();
    CHECK(!m.isOpen());
}

TEST(automap_reveals_lines_from_subsector) {
    Vertex v1{0, 0}, v2{10, 0};
    Sector sector{0, 0, "F", "C", 160, 0, 0};
    SideDef side{0, 0, "-", "-", "-", nullptr};
    LineDef line{nullptr, nullptr, 0, 0, 0, nullptr, nullptr};
    Seg seg = makeSegWithSecret(v1, v2, sector, side, line, false);
    Subsector subsector{{&seg}, seg.lineDef->frontSide->sector};

    AutoMap m;
    m.revealFromSubsector(&subsector);
    CHECK(std::find(m.revealedLines.begin(), m.revealedLines.end(), seg.lineDef) != m.revealedLines.end());
}

TEST(automap_secret_lines_not_revealed) {
    Vertex v1{0, 0}, v2{10, 0};
    Sector sector{0, 0, "F", "C", 160, 0, 0};
    SideDef side{0, 0, "-", "-", "-", nullptr};
    LineDef line{nullptr, nullptr, 0, 0, 0, nullptr, nullptr};
    Seg seg = makeSegWithSecret(v1, v2, sector, side, line, true);
    Subsector subsector{{&seg}, seg.lineDef->frontSide->sector};

    AutoMap m;
    m.revealFromSubsector(&subsector);
    CHECK(m.revealedLines.empty());
}

TEST(automap_zoom_clamped_to_min_max) {
    AutoMap m;
    m.panAndZoom(0, 0, -100);
    CHECK(m.zoom >= AUTOMAP_MIN_ZOOM);
    m.panAndZoom(0, 0, 100);
    CHECK(m.zoom <= AUTOMAP_MAX_ZOOM);
}

TEST(automap_pan_moves_center) {
    AutoMap m;
    m.panAndZoom(10, -5, 0);
    CHECK_EQ(m.centerX, 10.0f);
    CHECK_EQ(m.centerY, -5.0f);
}

TEST_MAIN()

File: tests/test_doom_automap.py
import unittest

from doom_automap import AutoMap
from doom_map import LineDef, Seg, Sector, SideDef, Subsector, Vertex


def _seg_with_secret(secret):
    sector = Sector(0, 0, "F", "C", 160, 0, 0)
    side = SideDef(0, 0, "-", "-", "-", sector)
    flags = LineDef.SECRET_FLAG if secret else 0
    line = LineDef(Vertex(0, 0), Vertex(10, 0), flags=flags, special=0, tag=0, frontSide=side)
    return Seg(line.v1, line.v2, 0, line, 0, 0)


class AutoMapToggleTest(unittest.TestCase):
    def test_toggle_flips_open_state(self):
        m = AutoMap()
        self.assertFalse(m.isOpen)
        m.toggle()
        self.assertTrue(m.isOpen)
        m.toggle()
        self.assertFalse(m.isOpen)


class RevealLinesTest(unittest.TestCase):
    """automap.md의 '## Feature: 지도에 선 드러내기'를 옮긴 테스트."""

    def test_reveals_lines_from_current_subsector(self):
        seg = _seg_with_secret(secret=False)
        subsector = Subsector(segs=[seg], sector=seg.lineDef.frontSide.sector)
        m = AutoMap()
        m.revealFromSubsector(subsector)
        self.assertIn(seg.lineDef, m.revealedLines)

    def test_secret_lines_are_not_revealed(self):
        seg = _seg_with_secret(secret=True)
        subsector = Subsector(segs=[seg], sector=seg.lineDef.frontSide.sector)
        m = AutoMap()
        m.revealFromSubsector(subsector)
        self.assertNotIn(seg.lineDef, m.revealedLines)

    def test_revealing_same_line_twice_does_not_duplicate(self):
        seg = _seg_with_secret(secret=False)
        subsector = Subsector(segs=[seg], sector=seg.lineDef.frontSide.sector)
        m = AutoMap()
        m.revealFromSubsector(subsector)
        m.revealFromSubsector(subsector)
        self.assertEqual(m.revealedLines.count(seg.lineDef), 1)


class PanZoomTest(unittest.TestCase):
    def test_zoom_is_clamped_to_min_max(self):
        m = AutoMap()
        m.pan_and_zoom(0, 0, zoomDelta=-100)
        self.assertGreaterEqual(m.zoom, 0.25)
        m.pan_and_zoom(0, 0, zoomDelta=100)
        self.assertLessEqual(m.zoom, 4.0)

    def test_pan_moves_center(self):
        m = AutoMap()
        m.pan_and_zoom(10, -5, zoomDelta=0)
        self.assertEqual((m.centerX, m.centerY), (10, -5))


if __name__ == "__main__":
    unittest.main()
